Builder completion phase should handle stage_and_commit as a direct mechanical step

## Problem

When the builder leaves uncommitted changes but hasn't pushed or created a PR, the shepherd's completion phase needs to perform `stage_and_commit` + `push_branch` + `create_pr`. However, `_direct_completion()` in `builder.py` only handles a subset of mechanical steps:

```python
mechanical_steps = {"push_branch", "add_review_label", "create_pr"}
if not steps or not set(steps).issubset(mechanical_steps):
    return False  # Falls through to agent-based completion
```

Since `stage_and_commit` is not in `mechanical_steps`, the entire completion falls back to spawning a full Claude agent with the builder role. This is unreliable because:

1. The agent is loaded with the full builder role (implementation instructions, testing, checkpoints, etc.) which conflicts with the simple "just commit and push" task
2. The completion args are passed as a long string via tmux send-keys, adding fragility
3. A 300s timeout Claude session for what should be a `git add -A && git commit` subprocess call is wasteful

This caused issue #2408's shepherd to fail — the builder correctly implemented the feature and tests passed, but the completion phase couldn't commit/push/create the PR in two attempts.

## Root Cause

`_direct_completion()` at `loom-tools/src/loom_tools/shepherd/phases/builder.py:~3037` explicitly excludes `stage_and_commit` from mechanical steps, even though it's equally mechanical: `git add <files> && git commit -m "<message>"`.

## Proposed Fix

Add `stage_and_commit` to the `_direct_completion()` method as a subprocess-based mechanical step:

1. Gather the list of modified/added files from the worktree
2. Run `git add` on the relevant files (not `-A` to avoid adding untracked junk like `node_modules`)
3. Run `git commit -m "feat: implement changes for issue #N"` (or derive a better message from the issue title)
4. Continue with the existing `push_branch` and `create_pr` direct completion logic

This eliminates the need to spawn a Claude agent for commit operations entirely.

## Files to Change

- `loom-tools/src/loom_tools/shepherd/phases/builder.py` — add `stage_and_commit` to `mechanical_steps` and implement the subprocess logic in `_direct_completion()`
- `loom-tools/tests/test_builder_phase.py` (or equivalent) — test that direct completion handles the stage+commit+push+PR flow

## Observed Failure

```
[04:33:16] [WARN] Builder left incomplete work (attempt 1/2): worktree exists (branch=feature/issue-2408, commits_ahead=0, uncommitted=2 files); remote branch missing; no PR
[04:33:17] [WARN] Completion phase failed with exit code 1
[04:33:42] [ERROR] builder phase validation failed: worktree exists (branch=feature/issue-2408, commits_ahead=0, uncommitted=2 files); remote branch missing; no PR
```

Both completion attempts failed to commit the builder's work, requiring manual intervention.

--- a/loom-tools/src/loom_tools/shepherd/phases/builder.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/builder.py
@@ -3156,36 +3156,10 @@ class BuilderPhase:
 
         for step in steps:
             if step == "stage_and_commit":
-                result = subprocess.run(
-                    ["git", "add", "-A"],
-                    cwd=ctx.worktree_path,
-                    capture_output=True,
-                    text=True,
-                    check=False,
-                )
-                if result.returncode != 0:
-                    log_warning(
-                        f"Direct completion: git add failed: "
-                        f"{result.stderr.strip()[:200]}"
-                    )
+                if not self._stage_and_commit(ctx):
+                    log_warning("Direct completion: stage_and_commit failed")
                     return False
-                result = subprocess.run(
-                    [
-                        "git", "commit", "-m",
-                        f"fix: implement changes for issue #{ctx.config.issue}",
-                    ],
-                    cwd=ctx.worktree_path,
-                    capture_output=True,
-                    text=True,
-                    check=False,
-                )
-                if result.returncode != 0:
-                    log_warning(
-                        f"Direct completion: git commit failed: "
-                        f"{result.stderr.strip()[:200]}"
-                    )
-                    return False
-                log_success("Direct completion: changes staged and committed")
+                log_success("Direct completion: changes committed")
 
             elif step == "push_branch":
                 if not self._push_branch(ctx):
@@ -3638,6 +3612,80 @@ class BuilderPhase:
         (e.g. after the doctor applies test fixes).
         """
         return self._push_branch(ctx)
+
+    def _stage_and_commit(self, ctx: ShepherdContext) -> bool:
+        """Stage meaningful changes and create a commit in the worktree.
+
+        Stages all non-artifact files (respects .gitignore) and commits with
+        a descriptive message.  Returns True if a commit was created.
+        """
+        if not ctx.worktree_path or not ctx.worktree_path.is_dir():
+            return False
+
+        # Get current status
+        status_result = subprocess.run(
+            ["git", "-C", str(ctx.worktree_path), "status", "--porcelain"],
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        if status_result.returncode != 0 or not status_result.stdout.strip():
+            log_warning("Direct completion: no changes to commit")
+            return False
+
+        # Split without stripping the full output — the leading status
+        # chars (e.g. " M") are significant in porcelain format.
+        porcelain_lines = [
+            line for line in status_result.stdout.splitlines() if line
+        ]
+        meaningful, _ = self._filter_build_artifacts(porcelain_lines)
+        if not meaningful:
+            log_warning(
+                "Direct completion: only build artifacts uncommitted, "
+                "nothing meaningful to commit"
+            )
+            return False
+
+        # Extract file paths from porcelain lines (format: "XY filename")
+        files_to_stage = []
+        for line in meaningful:
+            path = line[3:].strip().strip('"') if len(line) > 3 else line.strip()
+            if path:
+                files_to_stage.append(path)
+
+        if not files_to_stage:
+            return False
+
+        # Stage the files
+        stage_result = subprocess.run(
+            ["git", "-C", str(ctx.worktree_path), "add", "--"] + files_to_stage,
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        if stage_result.returncode != 0:
+            log_warning(
+                f"Direct completion: git add failed: "
+                f"{(stage_result.stderr or '').strip()[:200]}"
+            )
+            return False
+
+        # Commit
+        commit_msg = f"feat: implement changes for issue #{ctx.config.issue}"
+        commit_result = subprocess.run(
+            ["git", "-C", str(ctx.worktree_path), "commit", "-m", commit_msg],
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        if commit_result.returncode != 0:
+            log_warning(
+                f"Direct completion: git commit failed: "
+                f"{(commit_result.stderr or '').strip()[:200]}"
+            )
+            return False
+
+        return True
 
     def _push_branch(self, ctx: ShepherdContext) -> bool:
         """Push the current branch to remote.

--- a/loom-tools/tests/shepherd/test_phases.py
+++ b/loom-tools/tests/shepherd/test_phases.py
@@ -8182,23 +8182,31 @@ class TestBuilderDirectCompletion:
         assert "100" in call_args
         assert "--add-label" in call_args
 
-    def test_non_mechanical_steps_returns_false(self, mock_context: MagicMock) -> None:
-        """Should return False when non-mechanical steps remain."""
+    def test_stage_commit_push_create_pr(self, mock_context: MagicMock) -> None:
+        """Should handle full stage+commit+push+create_pr flow directly."""
         builder = BuilderPhase()
+        mock_context.repo_root = Path("/fake/repo")
+        mock_context.worktree_path = Path("/fake/worktree")
+        mock_context.config = ShepherdConfig(issue=42)
         diag = {
-            "has_uncommitted_changes": False,
-            "commits_ahead": 1,
+            "has_uncommitted_changes": True,
+            "commits_ahead": 0,
             "remote_branch_exists": False,
             "pr_number": None,
             "pr_has_review_label": False,
         }
-        # Inject a hypothetical non-mechanical step
-        with patch.object(
-            builder, "_diagnose_remaining_steps",
-            return_value=["resolve_conflicts", "push_branch"],
+
+        with (
+            patch.object(builder, "_stage_and_commit", return_value=True),
+            patch.object(builder, "_push_branch", return_value=True),
+            patch(
+                "loom_tools.shepherd.phases.builder.subprocess.run"
+            ) as mock_run,
         ):
+            mock_run.return_value = MagicMock(returncode=0, stderr="")
             result = builder._direct_completion(mock_context, diag)
-        assert result is False
+
+        assert result is True
 
     def test_push_failure_returns_false(self, mock_context: MagicMock) -> None:
         """Should return False when push fails."""
@@ -8348,114 +8356,142 @@ class TestBuilderDirectCompletion:
         call_args = mock_run.call_args[0][0]
         assert "feature/issue-55" in call_args
 
-    def test_stage_and_commit_full_pipeline(
+    def test_stage_and_commit_failure_returns_false(
         self, mock_context: MagicMock
     ) -> None:
-        """Should handle stage_and_commit + push + create_pr as mechanical steps."""
+        """Should return False when stage_and_commit fails."""
         builder = BuilderPhase()
         mock_context.repo_root = Path("/fake/repo")
         mock_context.worktree_path = Path("/fake/worktree")
+        mock_context.config = ShepherdConfig(issue=42)
         diag = {
             "has_uncommitted_changes": True,
             "commits_ahead": 0,
             "remote_branch_exists": False,
             "pr_number": None,
             "pr_has_review_label": False,
-            "branch": "feature/issue-42",
         }
 
-        with (
-            patch.object(builder, "_push_branch", return_value=True),
-            patch(
-                "loom_tools.shepherd.phases.builder.subprocess.run"
-            ) as mock_run,
-        ):
-            mock_run.return_value = MagicMock(returncode=0, stderr="")
+        with patch.object(builder, "_stage_and_commit", return_value=False):
             result = builder._direct_completion(mock_context, diag)
 
-        assert result is True
-        # Verify git add, git commit, and gh pr create were all called
-        calls = [c[0][0] for c in mock_run.call_args_list]
-        assert calls[0] == ["git", "add", "-A"]
-        assert calls[1][:2] == ["git", "commit"]
-        assert calls[2][:3] == ["gh", "pr", "create"]
+        assert result is False
 
-    def test_stage_and_commit_git_add_failure(
-        self, mock_context: MagicMock
+
+class TestBuilderStageAndCommit:
+    """Test _stage_and_commit for direct mechanical staging and committing."""
+
+    def test_stages_and_commits_meaningful_files(
+        self, mock_context: MagicMock, tmp_path: Path
     ) -> None:
-        """Should return False when git add fails."""
+        """Should stage meaningful files and create a commit."""
         builder = BuilderPhase()
-        mock_context.worktree_path = Path("/fake/worktree")
-        diag = {
-            "has_uncommitted_changes": True,
-            "commits_ahead": 0,
-            "remote_branch_exists": False,
-            "pr_number": None,
-            "pr_has_review_label": False,
-        }
+        mock_context.worktree_path = tmp_path
+        mock_context.config = ShepherdConfig(issue=42)
+
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run"
+        ) as mock_run:
+            # First call: git status --porcelain
+            status_result = MagicMock(
+                returncode=0, stdout=" M src/main.py\n M src/utils.py\n"
+            )
+            # Second call: git add
+            add_result = MagicMock(returncode=0, stderr="")
+            # Third call: git commit
+            commit_result = MagicMock(returncode=0, stderr="")
+            mock_run.side_effect = [status_result, add_result, commit_result]
+
+            result = builder._stage_and_commit(mock_context)
+
+        assert result is True
+        # Verify git add was called with the specific files
+        add_call = mock_run.call_args_list[1]
+        add_args = add_call[0][0]
+        assert add_args[:4] == ["git", "-C", str(tmp_path), "add"]
+        assert "src/main.py" in add_args
+        assert "src/utils.py" in add_args
+        # Verify commit message references the issue
+        commit_call = mock_run.call_args_list[2]
+        commit_args = commit_call[0][0]
+        assert "commit" in commit_args
+        assert any("issue #42" in arg for arg in commit_args)
+
+    def test_no_worktree_returns_false(self, mock_context: MagicMock) -> None:
+        """Should return False when worktree path is not set."""
+        builder = BuilderPhase()
+        mock_context.worktree_path = None
+        result = builder._stage_and_commit(mock_context)
+        assert result is False
+
+    def test_no_changes_returns_false(
+        self, mock_context: MagicMock, tmp_path: Path
+    ) -> None:
+        """Should return False when there are no changes to commit."""
+        builder = BuilderPhase()
+        mock_context.worktree_path = tmp_path
+
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run"
+        ) as mock_run:
+            mock_run.return_value = MagicMock(returncode=0, stdout="")
+            result = builder._stage_and_commit(mock_context)
+
+        assert result is False
+
+    def test_only_artifacts_returns_false(
+        self, mock_context: MagicMock, tmp_path: Path
+    ) -> None:
+        """Should return False when only build artifacts are uncommitted."""
+        builder = BuilderPhase()
+        mock_context.worktree_path = tmp_path
 
         with patch(
             "loom_tools.shepherd.phases.builder.subprocess.run"
         ) as mock_run:
             mock_run.return_value = MagicMock(
-                returncode=1, stderr="fatal: not a git repository"
+                returncode=0, stdout=" M Cargo.lock\n"
             )
-            result = builder._direct_completion(mock_context, diag)
+            result = builder._stage_and_commit(mock_context)
 
         assert result is False
 
-    def test_stage_and_commit_git_commit_failure(
-        self, mock_context: MagicMock
+    def test_git_add_failure_returns_false(
+        self, mock_context: MagicMock, tmp_path: Path
     ) -> None:
-        """Should return False when git commit fails."""
+        """Should return False when git add fails."""
         builder = BuilderPhase()
-        mock_context.worktree_path = Path("/fake/worktree")
-        diag = {
-            "has_uncommitted_changes": True,
-            "commits_ahead": 0,
-            "remote_branch_exists": False,
-            "pr_number": None,
-            "pr_has_review_label": False,
-        }
-
-        add_ok = MagicMock(returncode=0, stderr="")
-        commit_fail = MagicMock(returncode=1, stderr="nothing to commit")
+        mock_context.worktree_path = tmp_path
+        mock_context.config = ShepherdConfig(issue=42)
 
         with patch(
             "loom_tools.shepherd.phases.builder.subprocess.run"
         ) as mock_run:
-            mock_run.side_effect = [add_ok, commit_fail]
-            result = builder._direct_completion(mock_context, diag)
+            status_result = MagicMock(returncode=0, stdout=" M src/main.py\n")
+            add_result = MagicMock(returncode=1, stderr="error: could not add")
+            mock_run.side_effect = [status_result, add_result]
+            result = builder._stage_and_commit(mock_context)
 
         assert result is False
 
-    def test_stage_and_commit_allows_zero_commits_ahead_for_create_pr(
-        self, mock_context: MagicMock
+    def test_git_commit_failure_returns_false(
+        self, mock_context: MagicMock, tmp_path: Path
     ) -> None:
-        """create_pr safety guard should not block when stage_and_commit will create a commit."""
+        """Should return False when git commit fails."""
         builder = BuilderPhase()
-        mock_context.repo_root = Path("/fake/repo")
-        mock_context.worktree_path = Path("/fake/worktree")
-        diag = {
-            "has_uncommitted_changes": True,
-            "commits_ahead": 0,  # 0 now, but stage_and_commit will create one
-            "remote_branch_exists": False,
-            "pr_number": None,
-            "pr_has_review_label": False,
-            "branch": "feature/issue-42",
-        }
+        mock_context.worktree_path = tmp_path
+        mock_context.config = ShepherdConfig(issue=42)
 
-        with (
-            patch.object(builder, "_push_branch", return_value=True),
-            patch(
-                "loom_tools.shepherd.phases.builder.subprocess.run"
-            ) as mock_run,
-        ):
-            mock_run.return_value = MagicMock(returncode=0, stderr="")
-            result = builder._direct_completion(mock_context, diag)
+        with patch(
+            "loom_tools.shepherd.phases.builder.subprocess.run"
+        ) as mock_run:
+            status_result = MagicMock(returncode=0, stdout=" M src/main.py\n")
+            add_result = MagicMock(returncode=0, stderr="")
+            commit_result = MagicMock(returncode=1, stderr="nothing to commit")
+            mock_run.side_effect = [status_result, add_result, commit_result]
+            result = builder._stage_and_commit(mock_context)
 
-        # Should succeed — stage_and_commit makes the 0-commit guard inapplicable
-        assert result is True
+        assert result is False
 
 
 class TestBuilderCompletionRetryDefault:

